### DISCUSSION TITLE
Command Factory - Configuration CC set - sending partial parameters - format - AEOTEC Indoor Siren

### DISCUSSION THREAD

**Discussion by BJ205** [zwave-js/node-red-contrib-zwave-js#270] [Q&A]
I am 'playing' with node-red on my laptop before moving flows on to a raspberry pi.  Node-red-contrib-zwave-js installed flawlessly and I am happy working with binary sensors, can switch power sockets on and off, happy days.  Now I am working with an AEOTEC Indoor Siren, again, much success, I can play sounds, though the Siren doesn't accept the volume settings.  Using Command Factory I can ask for data, for instance from Parameter 2 as in the picture, and the Siren returns four messages because there are four partial parameters.
![Command Factory get](https://user-images.githubusercontent.com/129403203/228869504-2af13bc3-a631-4daf-8da5-a20d735c665a.JPG)
The bit that has me stumped is how to set singly or as a group, the four partial parameters that the Siren tells me about.  I read https://zwave-js.github.io/node-zwave-js/#/config-files/partial-parameters which gives an example of 7[0x01] as how to access partial parameters, and also the manual for the Siren which has an example of 0x01020304 to set the four partial parameters of a parameter.  I read about the Configuration CC set command Configuration CC 
```
SET
async set(
    options: ConfigurationCCAPISetOptions,
): Promise<SupervisionResult | undefined>;
```
but I can't find how to format the ConfigurationCCAPISetOptions, or any mention elsewhere about ConfigurationCCAPISetOptions.  I have tried everything I can think of in the Params section of the Command Factory, but I either get error messages, or nothing, or it just reports back four unchanged Partial Parameters.   Any help much appreciated.

**Comment by marcus-j-davies** [comment]
Hi @BJ205 

You can use the **CCAPI** for that - but it will require the integer that represents all 4 bits.
you are far better using the Value API - I tend to discourage the **CCAPI** these days 

In V9 that I am working on the **ValueAPI** is default.
here is how to do it via a function node

Setting a Value

```js
let targetValueID = {
}
let someValue = someValue
let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 12,
        "method": "setValue",
        "params": [targetValueID, someValue]
    }
}
return Message
```

Getting a Value

```js
let targetValueID = {
}
let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 12,
        "method": "getValue",
        "params": [targetValueID]
    }
}
return Message
```
See
https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/Value-API

To get the correct valueID - In the side panel UI - double click on a property, and you will get the ValueID

**Comment by marcus-j-davies** [comment]
If you still wanted to use the **CCAPI**.
your params will be. (but ValueAPI is far easier)

`bitMask` obviously not always needed

This is a  **ConfigurationCCAPISetOptions** payload 
` [ {"parameter": 2, bitMask: <BitMask>, "value": <Desired Value>, "valueSize": <Number>} ]`

**Reply by marcus-j-davies** [reply]
Sorry not sure If closed that in error or it was you - re-opened just in case

**Reply by BJ205** [reply]
Thank you for the extremely quick reply. 
You can't take away my CCAPI, I am only just getting comfortable with it :-)
I will have a go with both ValueAPI and CCAPI
Thanks again

**Reply by marcus-j-davies** [reply]
Don't worry, CCAP, I won't go anywhere, it just won't be the preferred choice

**Reply by marcus-j-davies** [reply]
Also, you can set configuration properties in the UI

**Reply by BJ205** [reply]
OK, I am probably being a bit thick here, but still banging my head against a brick wall :-)
I am trying to use the system by sending data from a Function Block instead of a CMD Factory block, but keep getting an error message.
From the top, I used your method to getDefinedValueIDs for a Fibaro Wall Switch, the first item in the array gives the On/Off state.
![List of ValueIds](https://user-images.githubusercontent.com/129403203/229133317-3a2a8866-3edf-406d-8f4c-b815ce4dc2ee.JPG)
Then I used Command Factory CCAPI Command, Binary Switch Get, and I can get the correct answer.
![Using Command Factory Binary Switch](https://user-images.githubusercontent.com/129403203/229133829-84af66f6-3d48-42df-b801-1e4515892266.JPG)
The I went to the UI side bar to get the valueId for the same value.
![valueId from Side Panel](https://user-images.githubusercontent.com/129403203/229134291-282ab08b-97f8-42c1-9f36-f37503eb3401.JPG)
Put that in to a Function block.
![valueIFD inside Function node](https://user-images.githubusercontent.com/129403203/229134626-7cbf367c-71b7-4d61-94ad-e57833db453d.JPG)
Triggered it and get an error.
![ZWave debug info](https://user-images.githubusercontent.com/129403203/229134945-f97ef92c-b45d-43b0-afc5-220c3f83de31.JPG)
I am probably doing something that is obviously wrong, but it isn't obvious to me :-)
The bit that looks dodgy to me is that the valueId includes the current value of the Binary Switch, but I wouldn't know which bits to leave out.
If there is any other data you need, let me know.
As always, your help is appreciated.

**Reply by marcus-j-davies** [reply]
Hi @BJ205 

The Value ID - is only the value ID part - when you double click it gives you ALL the meta data - in that is also the ValueID
and it will be `targetValue`

`currentValue `- Current (read only)
`targetValue` - Desired Value 



![image](https://user-images.githubusercontent.com/55892693/229139328-126c62a5-5f35-469f-944c-13e59c8bec81.png)

**Reply by marcus-j-davies** [reply]
```js
let MyVID = {
   commandClassName: "Binary Switch",
   commandClass: 37,
   endpoint: 0,
   property: "targetValue",
   propertyName: "targetValue"
}

let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 12,
        "method": "setValue",
        "params": [MyVID, true]
    }
}
return Message
```

**Reply by marcus-j-davies** [reply]
The reason I emphasis the use of the ValueAPI - is that API will keep the UI in sync.
The CCAPI is extremely low level, and wouldn't alway keep the UI/local data in sync - as its affectively avoiding the "eyes" of the driver, so the driver wont be aware its changed  - therefore the UI not reporting current values  

Plus it splits partial params into separate ValueIDs

**Reply by marcus-j-davies** [reply]
`currentValue` - if you are reading
`targetValue `- if you are changing.

`currentValue` becomes `targetValue`  once the change has been made successfully

**Reply by BJ205** [reply]
Thank you, I told you it would be something obvious :-)
I had tried getting rid of the "meta" and "Propertyid" bits but had always left the valueId and its braces in place.
Now to try setting values on the Indoor Siren

**Reply by marcus-j-davies** [reply]
I have only included below the minimum required fields (normally)
just double clicking the value (or more accurately - the property) in the UI will hint at any special value in its ValueID

You could `getValue` for the `targetValue` property of course  - but it wouldn't really make sense 😅 

the CMD Node does support the `ValueAPI `Also

```js
let MyVID = {
   commandClass: 37, /* Binary Switch */
   endpoint: 0, /* i.e a plug with 2 outlets */
}

/* To Set A Value */

MyVID.property = 'targetValue'

let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 12,
        "method": "setValue",
        "params": [MyVID, true]
    }
}
return Message

/* To Get A Value */

MyVID.property = 'currentValue'

let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 12,
        "method": "getValue",
        "params": [MyVID]
    }
}
return Message
```

**Reply by BJ205** [reply]
So I am just getting the hang of it, and then you show me a slightly different method.  You have to understand that your brain might run at supersonic, mine only just makes ModelT  :-)
I am now writing the bits to do the Setting of values for the AEOTEC Siren.  It has 8 End Points with 7 different values that can be set for each End Point.
The first bits were easy to do with SoundSwitch SetConfugration in CMD Factory, that is working nicely, it sets a Default Tone and Default Volume for any End Point.
The next step is to be able to set a Light Effect for any End Point using a Function block and your information above.  If you see a large amount of smoke far away on the horizon you will know that I got frustrated :-)
Thank you again, I look forward to V9 being released.

**Reply by marcus-j-davies** [reply]
You welcome 🤭 

Do remember  - you can set most of this  in the UI (and for most Command Classes) 😁 
The UI uses the Value API.

![Screenshot 2023-03-31 at 17 02 32](https://user-images.githubusercontent.com/55892693/229172007-2568ef2e-925e-49a8-b532-45d8a28fac4e.png)

There is a feature that you might have missed.
set a value in the UI - and you will see the command, that you can then use your self as you see fit 🤓 

![Screenshot 2023-03-31 at 17 05 56](https://user-images.githubusercontent.com/55892693/229172719-9a56f030-e121-4ef8-844e-6d3e720cb85c.png)

**Reply by BJ205** [reply]
Yesterday I did spot that you can change all the settings in the UI, but thank you for pointing it out and increasing my pain 😁.
Didn't know about the ability to see the command, I shall be making use of that,
In my 'playing' I have developed a dashboard made with FlexDash, so though I could set everything in the ZWave UI, my version is easier to use, or at least I hope it will be easier to use..  And anyway, you always learn more having to do, and understand stuff, yes, I am a 'silver lining' type person 😁.  All this started because the AEOTEC Siren doesn't accept the Volume level when you ask it to play a sound, not your problem, from what I can find it is a firmware issue.  But it meant I wanted to be able to set the Default Volume in the 8 Endpoints that the siren has.  And once I was looking at setting those, it seemed a shame not to also be able to set the Light Effects, fade up, time on, fade down, time off, and how long a sound plays, and if it repeats the sounds and if it repeats then the delay between sounds.  And then you look back and realise you are down the 'rabbit hole'.
